packages/react-reconciler/src/__tests__/useMutableSource-test.internal.js

```diff
@@ -0,0 +1,1317 @@
+          // At high priority, toggle y so that it reads from A instead of B.
+          // Simultaneously, mutate A.
+          mutateA('high-pri baz');
+          root.render(
+            <App
+              getSnapshotFirst={getSnapshotA}
+              getSnapshotSecond={getSnapshotA}
+            />,
+          );
+
+          // If this update were processed before the next mutation,
+          // it would be expected to yield "high-pri baz" and "high-pri baz".
+        });
+
+        // At lower priority, mutate A again.
+        // This happens to match the initial value of B.
+        mutateA('bar');
+
+        // When this update is processed,
+        // it is expected to yield "bar" and "bar".
+      });
+
+      // Check that we didn't commit any inconsistent states.
+      // The actual sequence of work will be:
+      // 1. React renders the high-pri update, sees a new getSnapshot, detects the source has been further mutated, and throws
+      // 2. React re-renders with all pending updates, including the second mutation, and renders "bar" and "bar".
+      expect(Scheduler).toHaveYielded(['x: bar, y: bar']);
+    });
+
+    if (__DEV__) {
+      describe('dev warnings', () => {
+        it('should warn if the subscribe function does not return an unsubscribe function', () => {
+          const source = createSource('one');
+          const mutableSource = createMutableSource(source);
+
+          const brokenSubscribe = () => {};
+
+          expect(() => {
+            act(() => {
+              ReactNoop.render(
+                <Component
+                  label="only"
+                  getSnapshot={defaultGetSnapshot}
+                  mutableSource={mutableSource}
+                  subscribe={brokenSubscribe}
+                />,
+              );
+            });
+          }).toWarnDev(
+            'Mutable source subscribe function must return an unsubscribe function.',
+          );
+        });
+
+        it('should error if multiple renderers of the same type use a mutable source at the same time', () => {
+          const source = createSource('one');
+          const mutableSource = createMutableSource(source);
+
+          act(() => {
+            // Start a render that uses the mutable source.
+            ReactNoop.render(
+              <>
+                <Component
+                  label="a"
+                  getSnapshot={defaultGetSnapshot}
+                  mutableSource={mutableSource}
+                  subscribe={defaultSubscribe}
+                />
+                <Component
+                  label="b"
+                  getSnapshot={defaultGetSnapshot}
+                  mutableSource={mutableSource}
+                  subscribe={defaultSubscribe}
+                />
+              </>,
+            );
+            expect(Scheduler).toFlushAndYieldThrough(['a:one']);
+
+            const PrevScheduler = Scheduler;
+
+            // Get a new copy of ReactNoop.
+            loadModules();
```

Comment:
Could use another renderer like ART, instead of loading noop twice, so you don't have to mutate the flags

Reply:
I don't think our tests allow multiple renderers to be used without resetting things anyway, because of how the host config stuff works. I think that's why we do the module resetting elsewhere. :)

Reply:
Ah I got mixed and thought you were changing the flags in between runs. Got it :D